File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
class Solution {
    public ListNode reverse(ListNode head) {
        ListNode prev = null, cur = head;
        while (cur != null) {
            ListNode front = cur.next;
            cur.next = prev;
            prev = cur;
            cur = front;
        }
        return prev;
    }
    
    public ListNode findKthNode(ListNode start, int k) {
        while (start != null && k > 1) {
            k--;
            start = start.next;
        }
        return start;
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode cur = head, prevLast = null;
        while (cur != null) {
            ListNode kth = findKthNode(cur, k);
            if (kth == null) {
                if (prevLast != null) {
                    prevLast.next = cur;
                }
                break;
            }
            ListNode nextNode = kth.next;
            kth.next = null;
            
            reverse(cur);
            if (cur == head) {
                head = kth;
            } else {
                prevLast.next = kth;
            }
            
            prevLast = cur;
            cur = nextNode;
        }
        return head;
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
class Solution {
public:
    ListNode* reverseLL(ListNode* cur) {
        ListNode* prev = nullptr;
        
        while (cur != nullptr) {
            ListNode* front = cur->next;
            cur->next = prev;
            prev = cur;
            cur = front;
        }
        
        return prev;
    }
    
    ListNode* find_kth_node(ListNode* node, int k) {
        k--;
        while (k) {
            node = node->next;
            if (node == nullptr) {
                return nullptr;
            }
            k--;
        }
        return node;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* cur = head;        
        
        ListNode* prev = nullptr;
        while (cur != nullptr) {
            ListNode* kth = find_kth_node(cur, k);
            
            if (kth == nullptr) {
                if (prev != nullptr) {
                    prev->next = cur;
                }
                break;
            }
            
            ListNode* next_node = kth->next;
            kth->next = nullptr;
            
            reverseLL(cur);
            
            if (head == cur) {
                head = kth;
            } else {
                prev->next = kth;
            }
            
            prev = cur;
            cur = next_node;
        }
        
        return head;
    }
};